src/webgpu/shader/execution/expression/call/builtin/textureDimensions.spec.ts

```diff
@@ -0,0 +1,497 @@
+  }
+  for (let i = 0; i < params.textureMipCount - params.baseMipLevel; i++) {
+    out.push(i);
+  }
+  return out;
+}
+
+/** @returns the GPUTextureViewDimensions to test for the format and number of samples */
+function viewDimensions(params: {
+  format: GPUTextureFormat;
+  samples?: number;
+}): readonly GPUTextureViewDimension[] {
+  if (params.samples !== undefined && params.samples !== 1) {
+    return ['2d'];
+  }
+  const info = kTextureFormatInfo[params.format];
+  if (
+    isCompressedTextureFormat(params.format) ||
+    info.depth !== undefined ||
+    info.stencil !== undefined
+  ) {
+    // https://www.w3.org/TR/webgpu/#dom-gputexturedimension-2d
+    // Only "2d" textures may have mipmaps, be multisampled, use a compressed or depth/stencil
+    // format, and be used as a render attachment.
+    return ['2d', '2d-array', 'cube', 'cube-array'];
+  }
+  return ['1d', '2d', '2d-array', '3d', 'cube', 'cube-array'];
+}
+
+/** @returns the GPUTextureDimension for the GPUTextureViewDimension */
+function textureDimensionsForViewDimensions(dim: GPUTextureViewDimension): GPUTextureDimension {
+  switch (dim) {
+    case '1d':
+      return '1d';
+    case '2d':
+    case '2d-array':
+    case 'cube':
+    case 'cube-array':
+      return '2d';
+    case '3d':
+      return '3d';
+  }
+}
+
+/** TestValues holds the texture size and expected return value of textureDimensions() */
+type TestValues = {
+  /** The size of the texture to create */
+  size: number[];
+  /** The expected result of calling textureDimensions() */
+  expected: number[];
+};
+
+/** @returns The TestValues to use for the given texture dimensions and format */
+function testValues(params: {
+  dimensions: GPUTextureViewDimension;
+  format: GPUTextureFormat;
+  baseMipLevel: number;
+  textureDimensionsLevel?: number;
+}): TestValues {
+  const formatInfo = kTextureFormatInfo[params.format];
+  const bw = formatInfo.blockWidth;
+  const bh = formatInfo.blockHeight;
+  let mip = params.baseMipLevel;
+  if (params.textureDimensionsLevel !== undefined) {
+    mip += params.textureDimensionsLevel;
+  }
+
+  switch (params.dimensions) {
+    case '1d': {
+      const w = bw * 8;
+      return { size: [w], expected: [w >>> mip] };
+    }
+    case '2d': {
+      const w = bw * 4;
+      const h = bh * 8;
+      return { size: [w, h], expected: [w >>> mip, h >>> mip] };
+    }
+    case '2d-array': {
+      const w = bw * 8;
+      const h = bh * 4;
+      return { size: [w, h, 4], expected: [w >>> mip, h >>> mip] };
+    }
+    case '3d': {
+      const w = bw * 8;
+      const h = bh * 4;
+      const d = 16;
+      return {
+        size: [w, h, d],
+        expected: [w >>> mip, h >>> mip, d >>> mip],
+      };
+    }
+    case 'cube': {
+      const l = bw * bh * 4;
+      return {
+        size: [l, l, 6],
+        expected: [l >>> mip, l >>> mip],
+      };
+    }
+    case 'cube-array': {
+      const l = bw * bh * 4;
+      return {
+        size: [l, l, 12],
```

Comment:
The 12 here is a non-trivial multiple of the 6 faces required for a cube map. Make that clear.

Reply:
Done.